geps/gep-2648/index.md

```diff
@@ -0,0 +1,415 @@
+plugin.
+
+### Conditions
+
+Implementations using Policy objects MUST include a `spec` and `status` stanza,
+and the `status` stanza MUST contain a `conditions` stanza, using the standard
+Condition format.
+
+Policy authors should consider namespacing the `conditions` stanza with a
+`controllerName`, as in Route status, if more than one implementation will be
+reconciling the Policy type.
+
+#### On `Policy` objects
+
+Each Direct Attached Policy MUST populate the `Accepted` condition and reasons
+as defined in the Go spec, a snapshot of which is included below. The canonical
+representation is in the actual Go files. (At the time of writing, this is in
+`apis/v1alpha2/policy_types.go`)
+
+```go
+// PolicyConditionType is a type of condition for a policy.
+type PolicyConditionType string
+
+// PolicyConditionReason is a reason for a policy condition.
+type PolicyConditionReason string
+
+const (
+  // PolicyConditionAccepted indicates whether the policy has been accepted or rejected
+  // by a targeted resource, and why.
+  //
+  // Possible reasons for this condition to be True are:
+  //
+  // * "Accepted"
+  //
+  // Possible reasons for this condition to be False are:
+  //
+  // * "Conflicted"
+  // * "Invalid"
+  // * "TargetNotFound"
+  //
+  PolicyConditionAccepted PolicyConditionType = "Accepted"
+
+  // PolicyReasonAccepted is used with the "Accepted" condition when the policy has been
+  // accepted by the targeted resource.
+  PolicyReasonAccepted PolicyConditionReason = "Accepted"
+
+  // PolicyReasonConflicted is used with the "Accepted" condition when the policy has not
+  // been accepted by a targeted resource because there is another policy that targets the same
+  // resource and a merge is not possible.
+  PolicyReasonConflicted PolicyConditionReason = "Conflicted"
+
+  // PolicyReasonInvalid is used with the "Accepted" condition when the policy is syntactically
+  // or semantically invalid.
+  PolicyReasonInvalid PolicyConditionReason = "Invalid"
+
+  // PolicyReasonTargetNotFound is used with the "Accepted" condition when the policy is attached to
+  // an invalid target resource
+  PolicyReasonTargetNotFound PolicyConditionReason = "TargetNotFound"
+)
+```
+
+#### On targeted resources
+
+Implementations that use Direct Policy objects SHOULD put a Condition into
+`status.Conditions` of any objects affected by a Direct Policy, if that field
+is present.
+
+If they do, that Condition MUST have a `type` ending in `PolicyAffected` (like
+`gateway.networking.k8s.io/PolicyAffected`),
+and have the optional `observedGeneration` field kept up to date when the `spec`
+of the Policy-attached object changes.
+
+Implementations SHOULD use their own unique domain prefix for this Condition
+`type` - it is recommended that implementations use the same domain as in the
+`controllerName` field on GatewayClass (or some other implementation-unique
+domain for implementations that do not use GatewayClass).
```

Comment:
If we had multiple controllers populating the same conditions list, wouldn't we risk a loop where each controller was trying to populate the "correct" set of conditions?

Reply:
I can imagine maybe an API or protocol that the primary gateway controller would commit to, and thus take responsibility for all updates. That not being the case, we may depend on the multiple controllers just to be good citizens, I guess. But I could be oversimplifying. Not sure if this would still hold for – say, let's exaggerate – dozens of controllers. At the very least, it can get a bit noisy.

Reply:
As @guicassolato said on another comment, the way this is _supposed_ to work is that controllers are supposed to update Conditions lists _additively_, putting back any Condition they don't understand or update as they found it. And because Conditions are a listTypeMap, the ordering shouldn't matter as far as the apiserver is concerned (I think, would need to check with apimachinery if that's actually the case).

Reply:
Interesting - that feels like something we should drop into implementation guidelines so we avoid unfortunate surprises. 

I know we put a lot of thought into using `controllerName` in our nested conditions on Routes and Policies, are we saying that in those cases an implementation like Kuadrant would create a separate `RouteParentStatus` entry with their controller name, even if they're not the ones directly implementing the parent? I think that idea generally makes sense to me, but it gets more confusing when you have resources that we historically thought of as having status that was owned by a single controller.

Reply:
I'll make a note to do an update to the implemntation guidelines later about this.

Reply:
> controllers are supposed to update Conditions lists additively, putting back any Condition they don't understand or update as they found it

This seems to require a controller to be aware of any condition it may have theoretically set in the past instead of assuming full ownership of the conditions and replacing with the current set of conditions it knows about. I generally would prefer some form of namespacing/nesting in status as broadly as possible that allows each controller to clearly own separate lists. This PR doesn't need to solve this, but it would be nice to have some kind of note somewhere in the GEP covering this as a follow up area and/or unresolved.

Reply:
It's actually part of the contract of upstream Conditions that controllers don't touch Conditions they don't know about - precisely to solve this sort of case. I agree that namespacing or nesting is better, but we can't always rely on that, especially if you're affecting some Core object that doesn't have a namespaced Conditions list.